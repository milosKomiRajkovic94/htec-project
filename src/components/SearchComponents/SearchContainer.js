import React, { useState, useEffect } from "react";

import DocumentTitle from "react-document-title";

import { useTranslation } from "react-i18next";

import { connect } from "react-redux";

import { bindActionCreators } from "redux";

import { searchNews } from "../../actions/searchActions";

import Loader from "../../utils/Loader";

import Article from "../../utils/Article";

import ArticleDetails from "../../utils/ArticleDetails";

import { setAreLanguagesDisabled } from "../../actions/configActions";

const SearchContainer = ({
  currentLanguage,
  searchNewsAction,
  searchedNews,
  searchedNewsLoadingFailed,
  searchedNewsAreLoading,
  setAreLanguagesDisabledAction,
}) => {
  const { t } = useTranslation();

  const [searchQuery, setSearchQuery] = useState("");
  const [currentLanguageLocally, setCurrentLanguageLocally] = useState("");
  const [isMorePressed, setIsMorePressed] = useState(false);
  const [singleArticleDetailsData, setSingleArticleDetailsData] = useState({});

  function handleInputChange(e) {
    setSearchQuery(e.target.value);
    searchNewsAction(e.target.value, currentLanguage);
  }

  function onMore(data) {
    setIsMorePressed(true);
    setAreLanguagesDisabledAction(true);
    setSingleArticleDetailsData(data);
  }

  function onBack() {
    setIsMorePressed(false);
    setAreLanguagesDisabledAction(false);
    setSingleArticleDetailsData({});
  }

  useEffect(() => {
    if (currentLanguage !== currentLanguageLocally) {
      setCurrentLanguageLocally(currentLanguage);
      searchNewsAction(searchQuery, currentLanguage);
    }
  }, [
    currentLanguage,
    currentLanguageLocally,
    setCurrentLanguageLocally,
    searchNewsAction,
    searchQuery,
  ]);

  if (isMorePressed) {
    return (
      <DocumentTitle title={`HTEC App - ${t("search")}`}>
        <div className="containerBelow">
          <ArticleDetails
            onBack={onBack}
            singleArticleData={singleArticleDetailsData}
          />
        </div>
      </DocumentTitle>
    );
  }

  return (
    <DocumentTitle title={`HTEC App - ${t("search")}`}>
      <div className="containerBelow">
        <div className="titleOfContainerBelow" data-testid="search">
          <li>{t("searchTopNewsByTheTerm")}</li>
        </div>
        <form className="searchForm">
          <input
            placeholder={`${t("searchTerm")}...`}
            value={searchQuery}
            onChange={handleInputChange}
            className="searchTermInputField"
          />
        </form>
        <div
          className={
            !searchedNewsLoadingFailed && !searchedNewsAreLoading
              ? "articlesContainer"
              : "displayNone"
          }
        >
          {searchedNews &&
            searchedNews.map((news) => {
              return (
                <Article
                  key={`topNews${currentLanguage}${news.source.name}${news.title}`}
                  singleArticleData={news}
                  onMore={onMore}
                />
              );
            })}
        </div>
        <Loader
          classNameOfWrapper={
            searchedNewsAreLoading ? "centered" : "displayNone"
          }
          text={t("loading")}
        />
        <div className={searchedNewsLoadingFailed ? "centered" : "displayNone"}>
          {t("failedLoading")}
        </div>
      </div>
    </DocumentTitle>
  );
};

const mapStateToProps = (state) => {
  return {
    currentLanguage: state.configReducer.currentLanguage,
    searchedNews: state.searchReducer.searchedNews,
    searchedNewsLoadingFailed: state.searchReducer.searchedNewsLoadingFailed,
    searchedNewsAreLoading: state.searchReducer.searchedNewsAreLoading,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    searchNewsAction: bindActionCreators(searchNews, dispatch),
    setAreLanguagesDisabledAction: bindActionCreators(
      setAreLanguagesDisabled,
      dispatch
    ),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchContainer);
