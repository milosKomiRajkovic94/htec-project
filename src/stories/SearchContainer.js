import React, { useState } from "react";

import "./search.css";

export const SearchContainer = () => {
  const [searchQuery, setSearchQuery] = useState("");

  function handleInputChange(e) {
    setSearchQuery(e.target.value);
  }

  return (
    <div className="containerBelow">
      <div className="titleOfContainerBelow" data-testid="search">
        <li>Search top news from Great Britain by term</li>
      </div>
      <form className="searchForm">
        <input
          placeholder="Search term..."
          value={searchQuery}
          onChange={handleInputChange}
          className="searchTermInputField"
        />
      </form>
    </div>
  );
};
