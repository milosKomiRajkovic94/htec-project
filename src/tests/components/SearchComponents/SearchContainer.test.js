import React from "react";

import ReactDOM from "react-dom";
import { Provider } from "react-redux";

import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import SearchContainer from "../../../components/SearchComponents/SearchContainer";

import configureStore from "../../../store";

const store = configureStore();

afterEach(cleanup);

it("renders without crashing withouth props", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <SearchContainer />
    </Provider>,
    div
  );
});

it("renders navigation buttons correctly", () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <SearchContainer />
    </Provider>
  );
  expect(getByTestId("search")).toHaveTextContent("searchTopNewsByTheTerm");
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <SearchContainer />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
