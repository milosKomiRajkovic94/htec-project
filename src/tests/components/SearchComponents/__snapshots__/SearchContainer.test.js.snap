// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="containerBelow"
>
  <div
    className="titleOfContainerBelow"
    data-testid="search"
  >
    <li>
      searchTopNewsByTheTerm
    </li>
  </div>
  <form
    className="searchForm"
  >
    <input
      className="searchTermInputField"
      onChange={[Function]}
      placeholder="searchTerm..."
      value=""
    />
  </form>
  <div
    className="displayNone"
  />
  <div
    className="loaderWrapperDefaultStyle centered "
  >
    <div
      className="css-0"
    >
      <div
        className="css-110do72"
      />
      <div
        className="css-1by9rry"
      />
      <div
        className="css-110do72"
      />
    </div>
    <p
      className="loaderTextDefaultStyle "
    >
      loading
    </p>
  </div>
  <div
    className="displayNone"
  >
    failedLoading
  </div>
</div>
`;
